core/src/main/java/forklift/consumer/MessageRunnable.java

```diff
@@ -42,51 +44,67 @@
     public void run() {
         RunAsClassLoader.run(classLoader, () -> {
             try {
-                try {
-                    // Validate the class.
-                    LifeCycleMonitors.call(ProcessStep.Validating, this);
-                    for (Method m : onValidate) {
-                        if (m.getReturnType() == List.class) {
-                            addError((List<String>)m.invoke(handler));
-                        } else if (m.getReturnType() == boolean.class) {
-                            error = error || !((boolean)m.invoke(handler));
-                        } else {
-                            addError("Return type of " + m.getReturnType() + " is not supported for OnValidate methods");
-                        }
+                // { Validating }
+                runHooks(ProcessStep.Validating);
+                for (Method m : onValidate) {
+                    if (m.getReturnType() == List.class) {
+                        addError((List<String>)m.invoke(handler));
+                    } else if (m.getReturnType() == boolean.class) {
+                        if (!(boolean)m.invoke(handler))
+                            addError("Validator " + m.getName() + " returned false");
+                    } else {
+                        throw new RuntimeException("onValidate method " + m.getName() + " has wrong return type " + m.getReturnType());
                     }
+                }
 
-                    // Run the message if there are no errors.
-                    if (error) {
-                        LifeCycleMonitors.call(ProcessStep.Invalid, this);
-                    } else {
-                        LifeCycleMonitors.call(ProcessStep.Processing, this);
-                        for (Method m : onMessage) {
-                            // Send the message to each handler.
-                            m.invoke(handler, new Object[] {});
-                        }
+                // Run the message if there are no errors.
+                if (errors.size() > 0) {
+                    // { Invalid }
+                    getErrors().stream().forEach(e -> log.error(e));
+                    runHooks(ProcessStep.Invalid);
+                    // !!EXIT!! if invalid, do not continue
+                    return;
```

Comment:
This will cause a message to not be ack'd and the queue will lock up. We need to ensure that any "handled" message is ack'd.